arbor/backends/gpu/stack.hpp

```diff
@@ -67,41 +81,67 @@ class stack {
         storage_(create_storage(capacity)), gpu_context_(gpu_ctx) {}
 
     ~stack() {
-        storage_.synchronize();
-        if (storage_->data) {
-            allocator<value_type>().deallocate(storage_->data, storage_->capacity);
+        memory::cuda_sync();
+        auto st = get_storage_copy();
+        if (st.data) {
+            allocator<value_type>().deallocate(st.data, st.capacity);
         }
     }
 
-    // Perform any required synchronization if concurrent host-side access is not supported.
-    // (Correctness still requires that GPU operations on this stack are complete.)
-    void host_access() const {
-        gpu_context_->synchronize_for_managed_access();
+    std::vector<T> get_data() const {
```

Comment:
If `data_` is updated by `update_host`, then this can return a `const std::vector<T>&`

Reply:
The copy of the data should be performed in `update_host`, and this function should just return a const reference.

Reply:
```C++
// give the device and host storage similar names (not host_copy_ and device_storage_)
storage_type* device_storage_;
storage_type host_storage_;

void update_host() {
    memory::cuda_memcpy_d2h(&host_storage_, device_storage_, sizeof(storage_type));
    const auto n = size();
    data_.resize(n);
    memory::cuda_memcpy_d2h(data_.data(), host_storage_.data, n*sizeof(T));
}

// replace get_data() with a much simpler data() call that returns a reference to whatever is on the host
const std::vector<T>& data() const {
    return data_;
}
```